common/src/main/java/io/pravega/common/util/ReusableFutureLatch.java

```diff
@@ -0,0 +1,210 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.util;
+
+import java.util.ArrayList;
+import java.util.concurrent.CompletableFuture;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * This class is similar to {@link ReusableLatch} but that works with {@link CompletableFuture} so
+ * that blocking can be async and exceptions and results can be passed.
+ * @param <T> The type of the futures that this class works with.
+ */
+public class ReusableFutureLatch<T> {
+    private final Object lock = new Object();
+    @GuardedBy("lock")
+    private final ArrayList<CompletableFuture<T>> waitingFutures = new ArrayList<>();
+    @GuardedBy("lock")
+    private T result;
+    @GuardedBy("lock")
+    private Throwable e;
+    @GuardedBy("lock")
+    private boolean released;
+    @GuardedBy("lock")
+    private Long runningThreadId;
+
+    public ReusableFutureLatch() {
+        released = false;
+    }
+
+    /**
+     * Supply a future to be notified when {@link #release(Object)} is called. If release has already been
+     * called, it will be compltedImmediatly.
+     * 
+     * @param toNotify The future that should be completed.
+     */
+    public void await(CompletableFuture<T> toNotify) {
```

Comment:
I wonder if `await` is the right name for this method. Just by looking at it, I'd expect the method to await the future (i.e., block sync until the future is done). Maybe rename it to `notify`?